### PR TITLE
drivers/ws281x: Add saul support

### PR DESCRIPTION
**Pull request by MrKevinWeiss**
### Contribution description

Support the ws281x driver on saul... starting with :drum: the `feather-nrf52840-sense` .

Note that this only allows control of all LEDs at once since saul does not really have a way to handle subindex or array control (for example LED array). For our use case we would like to keep it simple now but if we did want to allow saul to control individual LEDs in the future here are some considerations:
- `saul_entries` would be of size `WS281X_NUM * WS281X_PARAM_NUMOF` and one could iterate over this matrix, though there would still be no mechanism to address each one...
- The `ws281x_t` should change to be for each LED and not an array (likely a bit of rework) but then each `dev` could be for each LED
- Add a mechanism in saul to not only interact with an object, but an array of objects.

All of this is out of scope of this PR, but if people need individual control with saul, these issues must be considered.

### Testing procedure

Play around with:
```
BOARD=feather-nrf52840-sense make -C examples/saul flash term
```

### CODE REVIEW

**Comment by MrKevinWeiss** [comment]
I haven't tested yet but will do it tomorrow then make it a real PR

File: drivers/ws281x/ws281x_saul.c [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+static int set_rgb_led(const void *dev, const phydat_t *res)
+{
+    color_rgb_t color = {
+        .r = res->val[0],
+        .g = res->val[1],
+        .b = res->val[2]
+    };
+    ws281x_set((ws281x_t *)dev, 0, color);
+    return 1;
```
Comment (benpicco) [inline drivers/ws281x/ws281x_saul.c:35]: Don't you need a `ws281x_write()` to apply this?

**Comment by MrKevinWeiss** [comment]
There were a few bumps along the way... Since we have a whole bunch of backends in submodules we could not use the `driver_with_saul.mk` so I just copied an pasted the interesting bits (otherwise it would bring in all the backends).

Also, it seems there were some issues with the dep modelling, since xtimer is used in the header it should be mandatory for all backends...  I didn't see this until I added the `ws281x_write()` that @benpicco suggested.  Now I actually can see the colors.

I also accounted for the amount of LEDs a bit better... SAUL will only apply color to all LEDs and individual LEDs are not exposed.

File: drivers/ws281x/ws281x_saul.c [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+static int set_rgb_led(const void *dev, const phydat_t *res)
+{
+    ws281x_t *ws281x = (ws281x_t *)dev;
+    color_rgb_t color = {
+        .r = res->val[0],
+        .g = res->val[1],
+        .b = res->val[2]
+    };
+    for (int i = 0; i < ws281x->params.numof; i++) {
```
Comment (benpicco) [inline drivers/ws281x/ws281x_saul.c:35]: Here you set all LEDs, but in auto-init you create a SAUL entry for every individual LED

**Review by maribu — APPROVED** [review]
Thx :)

IMHO this is not scary and could be merged during soft feature freeze. But let's give @Teufelchen a chance to veto.

File: drivers/saul/init_devs/auto_init_ws281x.c [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+        if (ws281x_init(&ws281x_devs[i], &ws281x_params[i]) != 0) {
+            LOG_ERROR("ERROR\n");
+            continue;
+        }
+        LOG_DEBUG("SUCCESS\n");
+        saul_entries[i].dev = &(ws281x_devs[i]);
+        saul_entries[i].name = ws281x_saul_info[i].name;
+        saul_entries[i].driver = &ws281x_saul_driver;
+        saul_reg_add(&(saul_entries[i]));
```
Comment (Teufelchen1) [inline drivers/saul/init_devs/auto_init_ws281x.c:69]:
```suggestion
        saul_entries[i].dev = &(ws281x_devs[i]);
        saul_entries[i].name = ws281x_saul_info[i].name;
        saul_entries[i].driver = &ws281x_saul_driver;
        saul_reg_add(&(saul_entries[i]));
        LOG_DEBUG("SUCCESS\n");
```

**Review by Teufelchen1 — APPROVED** [review]
Approve, just a nit pick.

File: drivers/saul/init_devs/auto_init_ws281x.c (hunk 3 above)
Comment (MrKevinWeiss) [reply drivers/saul/init_devs/auto_init_ws281x.c:69]: Full disclosure, I copied this from other places... The `SUCCESS` message is actually for the initialization part, not the registry part.

**Comment by MrKevinWeiss** [comment]
> IMHO this is not scary and could be merged during soft feature freeze. But let's give @teufelchen a chance to veto.

I also don't mind waiting until the head freeze... Not so critical, just nice.

**Comment by maribu** [comment]
Please squash at will.

File: drivers/ws281x/ws281x_saul.c [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+{
+    ws281x_t *ws281x = (ws281x_t *)dev;
+    color_rgb_t color = {
+        .r = res->val[0],
+        .g = res->val[1],
+        .b = res->val[2]
+    };
+    for (unsigned idx = 0; idx < ws281x->params.numof; idx++) {
+        puts("Setting LED");
+        ws281x_set(ws281x, idx, color);
```
[the remark this answers is not in the document]
Comment (chrysn) [reply drivers/ws281x/ws281x_saul.c:37]: I think an extra vector component would be weird; strand-as-a-whole is at least consistent with the data types and write mode.

A per-LED mode would make sense, possibly also a group-of-LEDs mode – but maybe that can grow from this as a starting point?

**Comment by chrysn** [comment]
Build tests are blocked by the previously 3-year-dormant ws281x module coming back to life, as in [this build failure](https://ci.riot-os.org/details/43c082a382e14af790bb0c6c376c1ae6) -- apparently there was one of the non-inline-to-inline API changes that are nonbreaking for C but breaking for Rust. Looking into a quick fix.

File: drivers/ws281x/ws281x_saul.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (miri64) [reply drivers/ws281x/ws281x_saul.c:37]: > […] but maybe that can grow from this as a starting point?

Fine by me, just wanted to give a possible discussion starter. Maybe we should even distinguish from a single RGB LED and a LED strip in SAUL type. In any way it should be clarified (in the documentation) and unify that at the moment `SAUL_ACT_LED_RGB` sets just everything in the SAUL device to one color. Currently, it seems to be used for PWM-based RGB LEDs (where it is used to set the color of a single LED associated to a single PWM device) and by the Grove LED bar, that currently just seems to set the red compontent.

https://github.com/RIOT-OS/RIOT/blob/680c5aa23f84feba6999bf41f4642aef7825cb86/drivers/saul/pwm_saul.c#L102-L123

https://github.com/RIOT-OS/RIOT/blob/680c5aa23f84feba6999bf41f4642aef7825cb86/drivers/grove_ledbar/grove_ledbar_saul.c#L27-L32

[the remark this answers is not in the document]
Comment (chrysn) [reply drivers/ws281x/ws281x_saul.c:37]: And a good starter it is indeed.

I don't think the SAUL type is the right place. Single LED, LED strip or group of LEDs are just what RIOT sees, and may not correspond to the physical reality. (For example, on a PWM port there could be a strip of non-addressable LEDs, or all the many LEDs of a WS281x may be hidden behind a larger diffusor).

It is weird that in the Grove bar it uses one component -- anything monochromatic should be SAUL_ACT_DIMMER (and read one channel), and anything RGBish should be SAUL_ACT_LED_RGB and read the 3 components. (And I do not yet know how to best work with RGBW or RGBA LEDs in SAUL, but one step at a time).

**Comment by chrysn** [comment]
I have pushed a commit that updates the riot-wrappers version to include https://github.com/RIOT-OS/rust-riot-wrappers/pull/88 which fixes building when combining WS281x and Rust.

File: drivers/ws281x/ws281x_saul.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (MrKevinWeiss) [reply drivers/ws281x/ws281x_saul.c:37]: Can we say this is for a different PR? This issues is already documented in the PR description.

[the remark this answers is not in the document]
Comment (miri64) [reply drivers/ws281x/ws281x_saul.c:37]: Yes!

**Pull request merged**